caffe2/intra_op_parallel/intra_op_parallel.cc

```diff
@@ -0,0 +1,199 @@
+#include "intra_op_parallel.h"
+
+#include <functional>
+
+// #define INTRA_OP_PARALLEL_MEASURE_TIME_BREAKDOWN
+
+#ifdef INTRA_OP_PARALLEL_MEASURE_TIME_BREAKDOWN
+#include <chrono>
+#endif
+
+#ifdef CAFFE2_USE_MKL
+#include <mkl.h>
+#endif
+#ifdef CAFFE2_INTRA_OP_PARALLEL_CAN_USE_TBB
+#include <tbb/parallel_for.h>
+#endif
+
+C10_DEFINE_int(
+    caffe2_intra_op_parallel_max_num_tasks,
+    65536,
+    "The maximum number of tasks used for intra-op parallelism per operator");
+C10_DEFINE_int(
+    caffe2_intra_op_parallel_max_num_workers,
+    65536,
+    "The maximum number of tasks used for intra-op parallelism per operator. "
+    "Deprecated: please use caffe2_intra_op_parallel_max_num_tasks instead");
+C10_DEFINE_bool(
+    caffe2_intra_op_parallel_only_grab_idle_threads,
+    true,
+    "When true, only grab idle threads to minimize over-subscription");
+
+using namespace std;
+
+namespace caffe2 {
+
+namespace intra_op_parallel {
+
+ParallelOpBase::ParallelOpBase(const OperatorDef& operator_def, Workspace* ws)
+    : Operator<CPUContext>(operator_def, ws), count_(0) {
+#ifdef CAFFE2_INTRA_OP_PARALLEL_CAN_USE_TBB
+  use_tbb_ = operator_def.engine() == "TBB";
+#endif
+
+  max_num_tasks_ = std::max(
+      this->GetSingleArgument<int>("max_num_tasks", -1),
+      this->GetSingleArgument<int>("max_num_workers", -1));
+  if (max_num_tasks_ == -1) {
+    max_num_tasks_ = std::min(
+        FLAGS_caffe2_intra_op_parallel_max_num_tasks,
+        FLAGS_caffe2_intra_op_parallel_max_num_workers);
+  } else {
+    max_num_tasks_ = std::min(
+        max_num_tasks_,
+        std::min(
+            FLAGS_caffe2_intra_op_parallel_max_num_tasks,
+            FLAGS_caffe2_intra_op_parallel_max_num_workers));
+  }
+}
+
+ParallelOpBase::~ParallelOpBase() {}
+
+mutex& ParallelOpBase::TaskPoolMutex(int numa_node_id) {
+  // additional 1 for numa_node_id == -1
+  static vector<mutex> mutexes(std::max(0, GetNumNUMANodes()) + 1);
+  CAFFE_ENFORCE_GE(numa_node_id, -1);
+  CAFFE_ENFORCE_LT(numa_node_id, std::max(GetNumNUMANodes(), 0));
+  return mutexes[numa_node_id + 1];
+}
+
+bool ParallelOpBase::RunOnDevice() {
+#ifdef INTRA_OP_PARALLEL_MEASURE_TIME_BREAKDOWN
+  t_begin_ = chrono::system_clock::now();
+#endif
+
+  int num_tasks = 1;
+  ExecutorHelper* executor_helper = OperatorBase::GetExecutorHelper();
+  if (executor_helper) {
+    TaskThreadPoolBase* pool = executor_helper->GetPool(device_option());
+    assert(pool);
+
+    // In TBB we can't know the number of idle worker threads
+    num_tasks = std::min<int>(
+        max_num_tasks_,
+        FLAGS_caffe2_intra_op_parallel_only_grab_idle_threads && !use_tbb_
+            ? pool->numAvailable() + 1
+            : pool->size());
+  }
```

Comment:
So it means there would be oversubscription issue with TBB?

Reply:
It's the number of tasks/chunks, not the threads. TBB actually like when there are more tasks than CPUs, so it can balance load better. Oversubscription is bad with threads, which are managed by the TBB in a process-wide thread pool and reused between task arenas and algorithms.